src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -950,38 +958,46 @@ private ArrayTypeSymbol VisitArrayInitializer(BoundArrayCreation node)
             var arrayType = (ArrayTypeSymbol)node.Type;
             var elementType = arrayType.ElementType;
 
-            var elementBuilder = ArrayBuilder<BoundExpression>.GetInstance();
-            GetArrayElements(node.InitializerOpt, elementBuilder);
+            var initialization = node.InitializerOpt;
+            var elementBuilder = ArrayBuilder<BoundExpression>.GetInstance(initialization.Initializers.Length);
+            GetArrayElements(initialization, elementBuilder);
 
-            var typeBuilder = (node.Syntax.Kind() == SyntaxKind.ImplicitArrayCreationExpression) ? ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance(elementBuilder.Count) : null;
-            foreach (var element in elementBuilder)
+            int n = elementBuilder.Count;
+            var resultBuilder = ArrayBuilder<Result>.GetInstance(n);
+            for (int i = 0; i < n; i++)
             {
-                VisitRvalue(element);
-                Result elementResult = _result;
-                if (typeBuilder != null)
-                {
-                    typeBuilder.Add(elementResult.Type);
-                }
-                else if (elementType?.IsReferenceType == true)
+                var element = RemoveImplicitConversions(elementBuilder[i]);
+                elementBuilder[i] = element;
+                resultBuilder.Add(VisitRvalueWithResult(element));
+            }
+
+            if (node.Syntax.Kind() == SyntaxKind.ImplicitArrayCreationExpression)
```

Comment:
>node.Syntax.Kind() == SyntaxKind.ImplicitArrayCreationExpression [](start = 16, length = 64)

It feels like it would be better to have this information in the bound node, rather than trying to infer from syntax.

Reply:
Added `// PROTOTYPE` comment.

---
In reply to: [169856215](https://github.com/dotnet/roslyn/pull/24935#discussion_r169856215) [](ancestors = 169856215)